scapy/contrib/isotp.py

```diff
@@ -355,37 +401,45 @@ def push(self, piece):
                     isotp_data = "".join(map(str, self.pieces))
                 self.ready = isotp_data[:self.total_len]
 
-    def __init__(self, use_ext_addr=None, did=None, basecls=None):
+    def __init__(self, use_ext_addr=None, did=None, basecls=ISOTP):
+        # type: (Optional[bool], Optional[Union[int, List[int], Iterable[int]]], Type[Packet]) -> None  # noqa: E501
+        """Initialize a ISOTPMessageBuilder object
+
+        :param use_ext_addr: True for only attempting to defragment with
+                             extended addressing, False for only attempting
+                             to defragment without extended addressing,
+                             or None for both
+        :param did: Destination Identifier
+        :param basecls: The class of packets that will be returned,
+                        defaults to ISOTP
         """
-        Initialize a ISOTPMessageBuilder object
-
-        :param use_ext_addr:    True for only attempting to defragment with
-                                extended addressing, False for only attempting
-                                to defragment without extended addressing,
-                                or None for both
-        :param basecls:         the class of packets that will be returned,
-                                defaults to ISOTP
-
-        """
-        self.ready = []
-        self.buckets = {}
+        self.ready = []  # type: List[Tuple[int, Optional[int], ISOTPMessageBuilder.Bucket]]  # noqa: E501
+        self.buckets = {}  # type: Dict[Tuple[Optional[int], int, int], ISOTPMessageBuilder.Bucket]  # noqa: E501
         self.use_ext_addr = use_ext_addr
-        self.basecls = basecls or ISOTP
-        self.dst_ids = None
-        self.last_ff = None
-        self.last_ff_ex = None
+        self.basecls = basecls
+        self.dst_ids = None  # type: Optional[Iterable[int]]
+        self.last_ff = None  # type: Optional[Tuple[Optional[int], int, int]]
+        self.last_ff_ex = None  # type: Optional[Tuple[Optional[int], int, int]]  # noqa: E501
         if did is not None:
-            if hasattr(did, "__iter__"):
+            if isinstance(did, list):
                 self.dst_ids = did
-            else:
+            elif isinstance(did, int):
                 self.dst_ids = [did]
+            elif hasattr(did, "__iter__"):
+                self.dst_ids = did
+            else:
+                raise TypeError("Invalid type for argument did!")
 
     def feed(self, can):
+        # type: (Union[Iterable[Packet], Packet]) -> None
         """Attempt to feed an incoming CAN frame into the state machine"""
         if not isinstance(can, Packet) and hasattr(can, "__iter__"):
             for p in can:
                 self.feed(p)
             return
+
+        can = cast(Packet, can)
```

Comment:
My understands when you use direct ifs. For instance
```
if isinstance(can, Packet):
    identifier = can.identifier
elif hasattr(...):
    ...
```
shouldn't require a cast. However it fails to guess using `not` and `Union`
```